### ISSUE TITLE
`check` and `check --watch` have different results

### ISSUE THREAD

**Issue by fosskers** [astral-sh/ty#1973]
### Summary

Hi there, thanks for `ty`.

In experimenting with `--watch` today I noticed that it reports fewer results than a normal `ty check` does. I'm running it like:
```
uv run ty check FOO/ --watch
```
Currently, this reports no errors under `FOO/`. However without `--watch` it reports 9.

### Version

0.0.4

**Comment by zanieb** [comment]
Can you share a minimal example we can use to reproduce?

**Comment by fosskers** [comment]
Unfortunately the code is not open source.

**Comment by MichaReiser** [comment]
Would you be able to share the logs when running `ty check FOO -v` and `ty check FOO -v --watch`?

**Comment by zanieb** [comment]
Can you produce it with a minimized example of the code then?

**Comment by fosskers** [comment]
Caught in the wild: just `uv run ty check` yields 25 results, while `--watch` only detects 16.

I'm not seeing anything particularly incriminating in the output of `-v`, other than some files taking more than 100ms to check. Note though that `-v` with `--watch` doesn't really work, as `--watch` clears the screen.

**Comment by fosskers** [comment]
`--watch` just found 20 diagnostics, yet I changed absolutely nothing. Non-determinism? Race conditions in parallel checks? Timeouts?

I'm on `0.0.4` at the moment.

Update: It's back down to 16 ... now back up to 21.

**Comment by MichaReiser** [comment]
Do you see the difference on the first run of `ty check --watch` or are the diagnostics only different after you made changes (and check ran multiple times?)

**Comment by fosskers** [comment]
If I activate `ty check --watch`, kill it, then reactivate, then repeat, yes I see it alternating between 16 results and 20 results, with no other code changes from me.

**Comment by MichaReiser** [comment]
With kill and reactivate, you mean you stop ty, then re-run the command?

**Comment by fosskers** [comment]
Correct, via `Ctrl-C`.

**Comment by MichaReiser** [comment]
Very interesting:

* ty doesn't use any persistent caching. Each run is a fresh run
* The `--watch` and non-watch use the exact same code paths. The only difference is that `ty check` (without watch) exits the main loop immediately after check completed once

**Comment by fosskers** [comment]
Tried to kill my editor and run `ty` in isolation to see if my LSP was somehow `touch`ing files and confusing `--watch`, but no, same observed behaviour.

**Comment by fosskers** [comment]
```rust
    fn watch(mut self, db: &mut ProjectDatabase) -> Result<ExitStatus> {
        tracing::debug!("Starting watch mode");
        let sender = self.sender.clone();
        let watcher = watch::directory_watcher(move |event| {
            sender.send(MainLoopMessage::ApplyChanges(event)).unwrap();
        })?;

        self.watcher = Some(ProjectWatcher::new(watcher, db));
        self.run(db)?;

        Ok(ExitStatus::Success)
    }
```
Could multiple events be fighting to apply changes at the same time?

**Comment by MichaReiser** [comment]
> Could multiple events be fighting to apply changes at the same time?

That should only be an issue for subsequent runs of the watch, but not the initial run. Also, updates abort all pending checks and ty then restarts.

**Comment by MichaReiser** [comment]
I tried to reproduce the issue by running ty on `discord.py` in watch mode but it always reported the same number of diagnostics. 

But I put up https://github.com/astral-sh/ruff/pull/22078, which should hopefully make this easier to debug. 

By chance, do any of the flaky diagnostics say something about `Divergent`? There are some rare cases where we experience non-determinism in ty and maybe they only happen to show up in watch moide (for reasons?)

**Comment by MichaReiser** [comment]
I'll close this as I haven't been able to reproduce, which makes the issue non-actionable for us. Feel free to comment on this issue if you're still experiencing this problem or open a new issue.
